pkg/sessionctx/variable/sysvar.go

```diff
@@ -1681,7 +1681,15 @@ var defaultSysVars = []*SysVar{
 				}
 			}
 			return normalizedValue, nil
-		}},
+		},
+		SetSession: func(sv *SessionVars, s string) error {
```

Comment:
I've changed it to follow a similar pattern with other variables. Because accessing the variable directly (in compile stage) will cause recursive lock:

```
pkg/domain/sysvar_cache.go:120 domain.(*Domain).rebuildSysVarCache ??? <<<<<

pkg/domain/sysvar_cache.go:119 domain.(*Domain).rebuildSysVarCache ???

pkg/domain/sysvar_cache.go:49 domain.(*Domain).rebuildSysVarCacheIfNeeded ???

pkg/domain/sysvar_cache.go:73 domain.(*Domain).GetGlobalVar ???

pkg/session/session.go:1397 session.(*session).GetGlobalSysVar ???

pkg/sessionctx/variable/session.go:2503 variable.(*SessionVars).GetSessionOrGlobalSystemVar ???

pkg/expression/infer_pushdown.go:464 expression.NewPushDownContextFromSessionVars ???

pkg/planner/core/util.go:474 core.GetPushDownCtx ???

pkg/planner/core/exhaust_physical_plans.go:2608 core.(*LogicalProjection).exhaustPhysicalPlans ???

pkg/planner/core/find_best_task.go:579 core.(*baseLogicalPlan).findBestTask ???

pkg/planner/core/optimizer.go:1248 core.physicalOptimize ???

pkg/planner/core/optimizer.go:333 core.doOptimize ???

pkg/planner/core/optimizer.go:383 core.DoOptimize ???

pkg/planner/optimize.go:501 planner.optimize ???

pkg/planner/optimize.go:333 planner.Optimize ???

pkg/executor/compiler.go:103 executor.(*Compiler).Compile ???

pkg/session/session.go:2196 session.(*session).ExecuteStmt ???

pkg/session/session.go:2100 session.(*session).ExecuteInternalStmt ???

pkg/session/session.go:2072 session.(*session).ExecRestrictedSQL.func1 ???

pkg/session/session.go:2051 session.(*session).withRestrictedSQLExecutor ???

pkg/session/session.go:2055 session.(*session).ExecRestrictedSQL ???

pkg/domain/sysvar_cache.go:92 domain.(*Domain).fetchTableValues ???

pkg/domain/sysvar_cache.go:122 domain.(*Domain).rebuildSysVarCache ???

pkg/domain/sysvar_cache.go:49 domain.(*Domain).rebuildSysVarCacheIfNeeded ???

pkg/domain/sysvar_cache.go:73 domain.(*Domain).GetGlobalVar ???

pkg/session/session.go:1397 session.(*session).GetGlobalSysVar ???

pkg/sessionctx/variable/session.go:2503 variable.(*SessionVars).GetSessionOrGlobalSystemVar ???

pkg/expression/infer_pushdown.go:464 expression.NewPushDownContextFromSessionVars ???

pkg/planner/core/util.go:474 core.GetPushDownCtx ???

pkg/planner/core/exhaust_physical_plans.go:2608 core.(*LogicalProjection).exhaustPhysicalPlans ???

pkg/planner/core/find_best_task.go:579 core.(*baseLogicalPlan).findBestTask ???

pkg/planner/core/find_best_task.go:308 core.(*baseLogicalPlan).iteratePhysicalPlan ???

pkg/planner/core/find_best_task.go:233 core.(*baseLogicalPlan).enumeratePhysicalPlans4Task ???

pkg/planner/core/find_best_task.go:619 core.(*baseLogicalPlan).findBestTask ???

pkg/planner/core/optimizer.go:1248 core.physicalOptimize ???

pkg/planner/core/optimizer.go:333 core.doOptimize ???

pkg/planner/core/optimizer.go:383 core.DoOptimize ???

pkg/planner/optimize.go:501 planner.optimize ???

pkg/planner/optimize.go:333 planner.Optimize ???

pkg/executor/compiler.go:103 executor.(*Compiler).Compile ???

pkg/session/session.go:2196 session.(*session).ExecuteStmt ???

pkg/session/session.go:2100 session.(*session).ExecuteInternalStmt ???

pkg/session/session.go:2072 session.(*session).ExecRestrictedSQL.func1 ???

pkg/session/session.go:2051 session.(*session).withRestrictedSQLExecutor ???

pkg/session/session.go:2055 session.(*session).ExecRestrictedSQL ???

pkg/bindinfo/util.go:36 bindinfo.execRows ???

pkg/bindinfo/global_handle.go:205 bindinfo.(*globalBindingHandle).LoadFromStorageToCache.func1 ???

pkg/bindinfo/global_handle.go:672 bindinfo.(*globalBindingHandle).callWithSCtx ???

pkg/bindinfo/global_handle.go:204 bindinfo.(*globalBindingHandle).LoadFromStorageToCache ???

pkg/domain/domain.go:1876 domain.(*Domain).LoadBindInfoLoop ???


Previous place where the lock was grabbed (same goroutine)

pkg/domain/sysvar_cache.go:120 domain.(*Domain).rebuildSysVarCache ??? <<<<<

pkg/domain/sysvar_cache.go:119 domain.(*Domain).rebuildSysVarCache ???

pkg/domain/sysvar_cache.go:49 domain.(*Domain).rebuildSysVarCacheIfNeeded ???

pkg/domain/sysvar_cache.go:73 domain.(*Domain).GetGlobalVar ???

pkg/session/session.go:1397 session.(*session).GetGlobalSysVar ???

pkg/sessionctx/variable/session.go:2503 variable.(*SessionVars).GetSessionOrGlobalSystemVar ???

pkg/expression/infer_pushdown.go:464 expression.NewPushDownContextFromSessionVars ???

pkg/planner/core/util.go:474 core.GetPushDownCtx ???

pkg/planner/core/exhaust_physical_plans.go:2608 core.(*LogicalProjection).exhaustPhysicalPlans ???

pkg/planner/core/find_best_task.go:579 core.(*baseLogicalPlan).findBestTask ???

pkg/planner/core/find_best_task.go:308 core.(*baseLogicalPlan).iteratePhysicalPlan ???

pkg/planner/core/find_best_task.go:233 core.(*baseLogicalPlan).enumeratePhysicalPlans4Task ???

pkg/planner/core/find_best_task.go:619 core.(*baseLogicalPlan).findBestTask ???

pkg/planner/core/optimizer.go:1248 core.physicalOptimize ???

pkg/planner/core/optimizer.go:333 core.doOptimize ???

pkg/planner/core/optimizer.go:383 core.DoOptimize ???

pkg/planner/optimize.go:501 planner.optimize ???

pkg/planner/optimize.go:333 planner.Optimize ???

pkg/executor/compiler.go:103 executor.(*Compiler).Compile ???

pkg/session/session.go:2196 session.(*session).ExecuteStmt ???

pkg/session/session.go:2100 session.(*session).ExecuteInternalStmt ???

pkg/session/session.go:2072 session.(*session).ExecRestrictedSQL.func1 ???

pkg/session/session.go:2051 session.(*session).withRestrictedSQLExecutor ???

pkg/session/session.go:2055 session.(*session).ExecRestrictedSQL ???

pkg/bindinfo/util.go:36 bindinfo.execRows ???

pkg/bindinfo/global_handle.go:205 bindinfo.(*globalBindingHandle).LoadFromStorageToCache.func1 ???

pkg/bindinfo/global_handle.go:672 bindinfo.(*globalBindingHandle).callWithSCtx ???

pkg/bindinfo/global_handle.go:204 bindinfo.(*globalBindingHandle).LoadFromStorageToCache ???

pkg/domain/domain.go:1876 domain.(*Domain).LoadBindInfoLoop ???

pkg/session/session.go:3385 session.bootstrapSessionImpl ???

pkg/testkit/mockstore.go:294 testkit.bootstrap ???

pkg/session/session.go:3296 session.BootstrapSession ???

pkg/testkit/mockstore.go:264 testkit.CreateMockStoreAndDomain ???

pkg/testkit/mockstore.go:72 testkit.CreateMockStore ???

pkg/sessionctx/stmtctx/stmtctx_test.go:132 stmtctx_test.TestWeakConsistencyRead ???



```